### PR TITLE
Upgrade Finish Commit to handle a large number of files

### LINKED ISSUE #2575: Unable to commit a large commit.
I've added 210k files into the repo. 
It failed with message: `grpc: received message larger than max (67323059 vs. 20971520)`
I've executed `pachctl finish-commit <repo> <commit>`
it failed with the same message `grpc: received message larger than max (67323059 vs. 20971520)`

Looks like you need to adjust: "grpc.max_send_message_length" for grpc client.
Here is the same issues that has been solved using this method: 
https://github.com/grpc/grpc/issues/8098

### PR DESCRIPTION
**Pull request by sjezewski**
WIP

Fixes #2575 

This PR:

- upgrades pfs driver to use the new putFileRecords collection for any r/w to etcd for scratch data written during an open commit
- upgrades the collection to support:
    - a `ListPrefix` method
    - an upgrade to the `ListPrefix` iterator that will paginate query results from etcd

It's that last update that allows us to retrieve large queries from etcd without blowing past the grpc `MaxMsgSize` limit.

Remaining todos:

- clean up the redundant prefixes (scratch space and for the collection)
- implement the paginated queries for the iterator
- update the new test to just test `FinishCommit` limits
- benchmark the difference in timing

### CODE REVIEW

**Comment by sjezewski** [comment]
So this approach may be flawed. I'm going to write down what we've tried, what we know, what we suspect, and how we'll proceed

### Old Model

In the old model, we'd put a single document per `putfile` w a unique prefix. This would contain the list of content added, or a tombstone denoting a deletion.

Something like this

```
_, err = d.etcdClient.Put(ctx, path.Join(prefix, uuid.NewWithoutDashes()), tombstone)
```

And then at `FinishCommit` time, we'd read through these documents sorted in order by creation time to construct the final HashTree.

However, to solve #2575 we need to paginate the etcd query results (when getting / constructing the hashtree). The two mechanisms we have to do that are providing a `prefix` for the initial query, and then using the `from-key` or range for subsequent pages. However. We cannot specify a `from-key` and also sort by creation time. Observe:

```
18-01-03[11:29:33]:~:0$etcdctl --endpoints=http://127.0.0.1:32379 get --prefix pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/ --keys-only --limit=10 --sort-by="CREATE"
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-0/e16a5029c8c8491b882c26622e769645
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-1/adfbc5fc61b54d4fb4a661d343e18257
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-10/9a4ff9960a984af49f7cb2b8688cc871
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-100/31433b535c1f4a0296f991cbc256c2cd
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-101/aa0c5647643a45b6972402d8e74554ac
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-102/320ca1e96a0a4c0ebb48f7ffb50d5120
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-103/a9e52350c2a0459986ffb6b5884b74f8
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-104/0be81542c1f0491b9485e728a4a2c141
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-105/c56c43571bef40ceb6b55d8ce22f3f54
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-106/c21b7ddb808b4b50a1ac88f034dbb63f

18-01-03[11:29:39]:~:0$etcdctl --endpoints=http://127.0.0.1:32379 get --from-key pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-106/c21b7ddb808b4b50a1ac88f034dbb63f --keys-only --limit=10 --sort-by="CREATE"
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-11/7c132ca39d1449ef9be376cd5ee7abb7
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-106/c21b7ddb808b4b50a1ac88f034dbb63f
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-107/6c6ebe8bb90549c28ff30503b3d51515
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-108/1730f1369b0341248355a59323c32f34
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-109/1b5f7fc9589c40499760801c346f70a7
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-110/8fd6d1d205a7443181a6e92222fb6c60
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-111/34643bb879514180ba7d5d31d9c9f0ee
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-112/78564947e27240728abe79c28a40a6c6
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-113/f8c8bad758e745bfa52157b31f5e8300
pachyderm_pfs/scratch/TestUpdatePipelineStats_dataf5752fde9766/5e62973836af466a80ab997fc1b8125c/file-114/b5e00e60ec534f5eb31c75fc1cb98bf6
```

We should be seeing files `0` through `10` in the first batch, and `10` through `19` in the second. But we get lexigraphically sorted results. The sort happens after the filtering.

### New Model

So, the new model we opted for was instead of doing an insert of a document for each put, actually use etcd to do a transaction on the document (upsert) as needed.

So instead of the `put` above, we're doing something like:

```
	_, err = col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
 		recordsCol := d.putFileRecords.ReadWrite(stm)
 
 		var existingRecords pfs.PutFileRecords
 		err := recordsCol.Get(prefix, &existingRecords)
 		if err != nil && !col.IsErrNotFound(err) {
 			return err
 		}
 		if newRecords.Tombstone {
 			existingRecords.Tombstone = true
 			existingRecords.Records = nil
 		} else {
 			existingRecords.Split = newRecords.Split
 			existingRecords.Records = append(existingRecords.Records, newRecords.Records...)
 		}
 		// Now put the new data
 		recordsCol.Put(prefix, &existingRecords)
 		return nil
 	})
```
### What we see

This mostly works (the integration test suite mostly passed), and the newly added `TestManyFilesSingleCommit` passes, but the pfs test suite fails miserably.

In particular, the memory on my machine (32GB) gets eaten by etcd, at which point I see an error like:

```
2018-01-03T10:49:10-08:00 ERROR pfs.API.PutFile {"duration":267.096380606,"error":"etcdserver: request timed out","request":{"file":{"commit":{"repo":{"name":"test"},"id":"1a5956636a7945249a6af7c6e564f728"},"path":"foo"}}} 
2018-01-03T10:49:10-08:00 ERROR pfs.API.PutFile {"duration":23.390383328,"error":"etcdserver: request timed out","request":{"file":{"commit":{"repo":{"name":"test"},"id":"1a5956636a7945249a6af7c6e564f728"},"path":"foo"}}} 
panic: etcdserver: request timed out

goroutine 43394 [running]:
github.com/pachyderm/pachyderm/src/server/pfs/server.TestATonOfPuts.func1(0x1388, 0x64, 0xc4306b32e0, 0xc426178340, 0x1000992, 0x4, 0xc426beace0, 0x1039d02, 0x53, 0xc4270c7780)
	/home/sjezewski/go/src/github.com/pachyderm/pachyderm/src/server/pfs/server/server_test.go:2087 +0x249
created by github.com/pachyderm/pachyderm/src/server/pfs/server.TestATonOfPuts
	/home/sjezewski/go/src/github.com/pachyderm/pachyderm/src/server/pfs/server/server_test.go:2083 +0x2ed
FAIL	github.com/pachyderm/pachyderm/src/server/pfs/server	435.119s
Makefile:335: recipe for target 'test-pfs' failed
make: *** [test-pfs] Error 1

```

And the etcd logs:

```
2018-01-03 18:49:10.108882 W | etcdserver: apply entries took too long [6.849775633s for 16 entries]
2018-01-03 18:49:10.121605 W | etcdserver: avoid queries with large range/delete range!
2018-01-03 18:49:10.152136 I | etcdserver/api/v3rpc: grpc: Server.processUnaryRPC failed to write status: stream error: code = Canceled desc = "context canceled"
2018-01-03 18:49:10.154867 I | etcdserver/api/v3rpc: grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
```

(there are many more transport closing messages)

Now I've also modified the page size (the number of results etcd returns). While having it at 10k makes the performance of `FinishCommit` much better (~3s) in the `TestManyFilesSingleCommit` test (as opposed to 100, which makes that operation take 3m!). I tuned it back down to 100 to see if that helped the memory load on etcd. It made its memory footprint grow more slowly, but it did eventually consume all the memory and then fail as above.

### What we suspect

Some combination of:

- using the STM instead of the simple put creates too much overhead
- we're doing too many transactions within etcd at once

Is causing the memory usage to blow up on etcd. 

### How we'll proceed

- I'm going to instrument the old method of putting vs the new STM
- I'm also going to tune the client semaphore down to see if that helps this test `TestATonOfPuts` succeed at all (even if it takes longer). In this case we suspect that etcd doesn't handle many concurrent transactions well ... in which case we might want to implement a semaphore at the STM layer to meter etcd transactions
- I'll also try upgrading our image of etcd to see if a newer version ameliorates this issue

Based on these results, we may throw out this model and try a different one. *We have shied away from doing upsert on file infos in the past, and this performance bottleneck may be why*. If so, it's important to understand this going forward. This is my attempt to record this so we remember 'why' we made this choice.

### Different Model

e.g we could instead write many documents to etcd, but use a counter and include that in their key so that we don't need to sort the results ... the lexigraphically sorted results are also the results sorted by creation time

**Comment by sjezewski** [comment]
We're still using the upsert model, and tracked down some of the mystery outlined above.

Basically, [I took etcd's benchmark tool](https://github.com/coreos/etcd/blob/master/tools/benchmark/cmd/stm.go), modified it to:

- use our collection library
- add a new `putFileRecord` each time it was run

And this way I was able to reproduce the very low rate we were seeing (~12req/s).

With this tool I was also able to reproduce the results from this [etcd blog post](https://coreos.com/blog/transactional-memory-with-etcd3.html) which outlines what we should see in cases where there is high and low key contention.

Basically, our slowdown was caused by larger and larger payloads being written. If we modify the key N times and each time add a record, we're basically transferring `O(N^2)` size in total (sum of N things ... is N(N+1)/2). Using our collection lib alone in their benchmark tool didn't replicate our slow request rate, but adding something to the record each time matched our rate pretty closely.

So, the result was that we tuned down that `TestATonOfPuts` test so that it runs in a reasonable amount of time. The test still goes beyond the `collection.QueryPaginationLimit`, so we verify that we _can_ handle larger numbers of files without making the test take forever. This resolves issue #2575 

Additionally, I added a few benchmarks (since that `TestATonOfPuts` was reporting timing info for putting the files and finishing the commit) so that its easier to analyze the performance of those components in the future.

Also note -- etcd can still gobble up memory ... but it makes sense because of the data being transferred. The fact that the mem (according to the OS) was still being used by etcd is a function of go's GC, not etcd's active usage. (Still a bit annoying if you're running these locally, like me, but not a dealbreaker). But if we see etcd evictions / OOMs, this may be a pattern as to why. (Too much load too quickly).

**Comment by sjezewski** [comment]
Here are the results of the benchmarks as well:

```
18-01-08[17:27:52]:pachyderm:1$1$go test -v ./src/server/ -bench=BenchmarkPutManyFiles -run=BenchmarkPutManyFilesSingleCommit -timeout=3000s -benchtime 100ms
goos: linux
goarch: amd64
pkg: github.com/pachyderm/pachyderm/src/server
BenchmarkPutManyFilesSingleCommitFinishCommit1-8                2	  51069982 ns/op
BenchmarkPutManyFilesSingleCommitFinishCommit10-8    	       1	 184704487 ns/op
BenchmarkPutManyFilesSingleCommitFinishCommit100-8   	       1	1917883609 ns/op
BenchmarkPutManyFilesSingleCommit1-8                 	       1	8932577530 ns/op
BenchmarkPutManyFilesSingleCommit10-8                	       1	96257950347 ns/op
BenchmarkPutManyFilesSingleCommit100-8               	       1	965105909386 ns/op
PASS
ok  	github.com/pachyderm/pachyderm/src/server	2136.295s
```

Translating that into human durations ...

Finish Commit:

| # of Files | time/op |
| ---| ---|
| 1000 | 51ms/op|
|10000|184ms/op|
|100000|1.91s/op|

Putting Many Files:

| # of Files | time/op|
| 1000 | 8.93s/op|
| 10000 | 1m36s/op|
|100000|16m5s/op|

File: src/server/pfs/server/driver.go [hunk 1]
```diff
@@ -1413,7 +1398,7 @@ func (d *driver) filePathFromEtcdPath(etcdPath string) string {
 	split := strings.Split(trimmed, "/")
 	// we only want /path/to/file so we use index 3 (note that there's an "" at
 	// the beginning of the slice because of the lead /)
-	return path.Join(split[3:]...)
+	return path.Join(split[6:]...)
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: I'm assuming this changed from `3` to `6` because of the addition of collections. Could you update to comment to explain why?

File: src/server/pfs/server/driver.go [hunk 2]
```diff
@@ -1747,25 +1692,48 @@ func (d *driver) getTreeForFile(ctx context.Context, file *pfs.File) (hashtree.H
...
 	}
-	openTree := parentTree.Open()
-	if err := d.applyWrites(resp, openTree); err != nil {
-		return nil, err
-	}
-	tree, err := openTree.Finish()
+	return d.constructTreeFromPrefix(ctx, prefix, parentTree)
+}
+
+func (d *driver) constructTreeFromPrefix(ctx context.Context, prefix string, parentTree hashtree.HashTree) (hashtree.HashTree, error) {
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: This is a bit pedantic but I find the name `constructTreeFromPrefix` a little incongruous with the other names in this file. For one thing I think you should use `get` in place of `construct` since they have about the same connotation and `get` is what other similar functions use. Another thing is that it doesn't quite feel like it's constructing the tree "from" a prefix. Maybe `getTreeForPrefix`, how would you feel about that?

File: src/server/pfs/server/driver.go [hunk 3]
```diff
@@ -1950,12 +1918,7 @@ func (d *driver) deleteFile(ctx context.Context, file *pfs.File) error {
...
 	}
 
-	prefix, err := d.scratchFilePrefix(ctx, file)
-	if err != nil {
-		return err
-	}
-
-	_, err = d.etcdClient.Put(ctx, path.Join(prefix, uuid.NewWithoutDashes()), tombstone)
+	err = d.upsertPutFileRecords(ctx, file, &pfs.PutFileRecords{Tombstone: true})
 	return err
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: Merge this with the line above into a single return.

File: src/server/pfs/server/driver.go [hunk 4]
```diff
@@ -1972,76 +1935,122 @@ func (d *driver) deleteAll(ctx context.Context) error {
...
+	prefix, err := d.scratchFilePrefix(ctx, file)
+	if err != nil {
+		return err
+	}
+
+	_, err = col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
+		revision := stm.Rev(d.openCommits.Path(file.Commit.ID))
+		if revision == 0 {
+			return fmt.Errorf("commit %v is not open", file.Commit.ID)
+		}
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: Did you check to make sure that the first write to the system doesn't get recorded as revision `0`?

File: src/server/pfs/server/driver.go [hunk 5]
```diff
@@ -1972,76 +1935,122 @@ func (d *driver) deleteAll(ctx context.Context) error {
...
+			if err != nil {
+				return fmt.Errorf("error parsing filename %s as int, this likely means you're "+
+					"using split on a directory which contains other data that wasn't put with split",
+					path.Base(nodes[len(nodes)-1].Name))
+			}
+			indexOffset++ // start writing to the file after the last file
+		}
+		for i, record := range records.Records {
+			if err := tree.PutFile(path.Join(filePath, fmt.Sprintf(splitSuffixFmt, i+int(indexOffset))), []*pfs.Object{{Hash: record.ObjectHash}}, record.SizeBytes); err != nil {
+				return err
```
Comment (jdoliner) [inline src/server/pfs/server/driver.go]: The diff on this last part is a little bit hard to follow but I think it's almost all superficial changes that just come from the fact that you are now handling multiple write records and thus need to put the code in a loop... which leads to indentation... which leads to hairy diffs. Is that accurate?

File: src/server/pkg/collection/collection.go [hunk 6]
```diff
@@ -397,6 +406,33 @@ func (c *readonlyCollection) GetBlock(key string, val proto.Unmarshaler) error {
 	}
 }
 
+func endKeyFromPrefix(prefix string) string {
+	// Lexigraphically increment the last character
+	return prefix[0:len(prefix)-1] + string(byte(prefix[len(prefix)-1])+1)
```
Comment (jdoliner) [inline src/server/pkg/collection/collection.go]: What happens if the last character is the maximum value here so incrementing it leads to wrap around behavior? Is that something we need to worry about?

File: src/server/pkg/collection/collection.go [hunk 7]
```diff
@@ -249,6 +254,10 @@ func (c *readWriteCollection) DeleteAll() {
 	c.stm.DelAll(c.prefix)
 }
 
+func (c *readWriteCollection) DeleteAllWithPrefix(prefix string) {
+	c.stm.DelAll(path.Join(c.prefix, prefix) + "/")
+}
```
Comment (jdoliner) [inline src/server/pkg/collection/collection.go]: I'd remove the `With` from this method and `ListWithPrefix` I don't think the `With` is adding a ton of clarity.

File: src/server/pkg/collection/collection.go [hunk 8]
```diff
@@ -438,6 +482,38 @@ func (i *iterator) Next(key *string, val proto.Unmarshaler) (ok bool, retErr err
...
+	// Reached end of resp, try for another page
+	fromKey := string(i.resp.Kvs[len(i.resp.Kvs)-1].Key)
+	resp, err := i.etcdClient.Get(i.ctx, fromKey, etcd.WithRange(i.endKey), etcd.WithLimit(QueryPaginationLimit))
+	if err != nil {
+		return false, err
+	}
+	if len(resp.Kvs) == 1 {
+		// Only got the from key, there are no more kvs to fetch from etcd
+		return false, nil
+	}
```
Comment (jdoliner) [inline src/server/pkg/collection/collection.go]: This case seems a bit weird to me. Why does it return a single KV if there are no more Kvs left and shouldn't we do something with that single KV it returned?

File: src/server/pkg/collection/collection.go [hunk 9]
```diff
@@ -438,6 +482,38 @@ func (i *iterator) Next(key *string, val proto.Unmarshaler) (ok bool, retErr err
...
+	i.index = 1 // Move past the from key
```
Comment (jdoliner) [inline src/server/pkg/collection/collection.go]: This seems a bit weird to. Is it that the first key is going to always be the `from` key which we've already returned? Is this true even of the first request we send out?

File: src/server/pkg/pfsdb/pfsdb.go [hunk 10]
```diff
@@ -46,6 +47,17 @@ func RepoRefCounts(etcdClient *etcd.Client, etcdPrefix string) col.Collection {
...
 
+// PutFileRecords returns a collection of putFileRecords
+func PutFileRecords(etcdClient *etcd.Client, etcdPrefix string) col.Collection {
+	return col.NewCollection(
+		etcdClient,
+		path.Join(etcdPrefix, putFileRecordsPrefix),
+		nil,
+		nil,
+		nil,
+	)
```
Comment (jdoliner) [inline src/server/pkg/pfsdb/pfsdb.go]: I'd recommend adding a template to this even though you don't use indexes. In the commit invariants branch that will be merged in soon we use the template to make sure that you don't accidentally put one type in the collection and read out a different one. It's just an added safety check that I think is nice to have.

File: src/server/pfs/server/driver.go (hunk 1 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: That's correct. I will update the comment. Actually while I'm at it ... the existing scratch prefix puts some redundant namespacing in there (something like `/pachyderm_pfs/scratch`) that we don't need now that these keys are namespaced because of the collection.

File: src/server/pfs/server/driver.go (hunk 2 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Sounds good :+1:

File: src/server/pfs/server/driver.go (hunk 5 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Close. If you look at line 2010, I'm actually removing a level of indentation. That's because a single `putFileRecords` can have a `tombstone` _and_ have data to be written (whereas before it was an if/else). But yea ... didn't really change anything in here except for the indentation.

File: src/server/pkg/collection/collection.go (hunk 7 above)
Comment (sjezewski) [reply src/server/pkg/collection/collection.go]: Sounds good :+1:

File: src/server/pkg/collection/collection.go (hunk 8 above)
Comment (sjezewski) [reply src/server/pkg/collection/collection.go]: Yea, so the way the `fromKey` works is that when you specify that in a query, it returns that key as well. So if the previous 'page' ended on the key that is actually the last key, the subsequent query will return 1 result.

File: src/server/pkg/collection/collection.go (hunk 9 above)
Comment (sjezewski) [reply src/server/pkg/collection/collection.go]: Similar deal ... because etcd returns the `fromKey` in the results (and the previous page has already provided that data through the iterator) we want to move past that key in the results we return via the iterator.

File: src/server/pfs/server/driver.go (hunk 4 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: I did not. I think this is analogous to the previous incantation of the code, [seen here](https://github.com/pachyderm/pachyderm/pull/2599/files/f1cccb143f94c3b57ab97dd0e3e5bc90a39d5eb4#diff-4448519cd42a9ad4ceeaeb1804180343L1472)

But I can double check this.

File: src/server/pkg/collection/collection.go (hunk 6 above)
Comment (sjezewski) [reply src/server/pkg/collection/collection.go]: Realistically, I think we're ok.

Since go uses UTF-8 the maximum value would be the `\xff` value (which is a non printable character).

As long as users name their files sane things (within the ASCII range), even if they use the max printable ASCII character (which is `~` or `\x7e`), incrementing this value is a valid byte that will serve as a valid endKey.

There is however, the matter of what a user chooses to name a file. Right now we only make sure the null character isn't in the file path. So in theory a user could use non printable UTF-8 bytes in filenames, and if they supplied `\xff` then yea, this would overflow and cause a problem. [We should probably just validate file paths to be printable ASCII characters](https://github.com/pachyderm/pachyderm/issues/2614)

File: src/server/pfs/server/driver.go (hunk 4 above)
Comment (sjezewski) [reply src/server/pfs/server/driver.go]: Capturing our discussion offline ...

Reasoning about this code is tricky. From manual testing w etcdctl its hard to check a revision of a key. So instead we opted to check that the value here is non empty and the commit ID matches the current commit. Really that's the simplest check we can think of to guarantee the value exists (and isn't an empty struct unmarshalled from an empty string).

**Comment by sjezewski** [comment]
Ok ... I made all the changes above and tests are passing. @jdoliner if you want to take another look.

File: src/server/pkg/collection/collection.go (hunk 6 above)
Comment (jdoliner) [reply src/server/pkg/collection/collection.go]: This seems reasonable.

File: src/server/pkg/collection/collection.go (hunk 8 above)
Comment (jdoliner) [reply src/server/pkg/collection/collection.go]: I see, and this won't trip us up if there really only is `1` total key because we won't even get to this. That call to `Get` is above.
